staging/src/k8s.io/sample-controller/integration_test.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	extensionclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	exttesting "k8s.io/apiextensions-apiserver/test/integration/testserver"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	apitesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
```

Comment:
I think we need to move these two imports:

```
	apitesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
	"k8s.io/kubernetes/test/integration/framework"
```

into some generic, exported package from k/k. As far as I'm aware, none of the published repos depend upon anything in k8s.io/kubernetes at the moment.

Could we move `k8s.io/kubernetes/cmd/kube-apiserver/app/testing` into `k8s.io/apiserver/pkg/testing` perhaps?

I'm not too sure on the best place for `k8s.io/kubernetes/test/integration/framework`. Is there some sensible, shared place this could go @sttts? These could definitely be generally applicable/useful to external projects. Given this is actually some fairly complex testing code for etcd, it'd be best if we don't encourage users to 'vendor' it themselves (i.e. copy and paste the package), as there could well be changes to this code required in future versions of k8s that a user may depend on.

Reply:
We can't move them as long as they are part of k/k. There is discussion about creating k/kube-apiserver. Then using these would be possible. For now the only way to test a controller is to use a fake client.

Reply:
See
 * Proposal for test frameworks repo (https://github.com/kubernetes/community/pull/1524) and 
 * https://github.com/kubernetes-sig-testing/frameworks.

Also note that the same (or similar) comment was made in:
 * https://github.com/kubernetes/kubernetes/pull/50144#issuecomment-333955979